### ISSUE TITLE
weed backup - does not do backup

### ISSUE THREAD

**Issue by tadwiew** [seaweedfs/seaweedfs#4371]
Hi,

I am trying to do a backup of weed volumes - before massive files removal - but it looks that "weed backup" just  creates empty data/index files. Why ?

MY environment : 
[root@media4 weed-backup]# /opt/weed/bin/weed version
version 30GB 3.19 03c6f978f42a55810441c7a5a6c77da567bc99b7 linux amd64

I have 1 Master server and 2 weed volume servers . All on CentOS 6.x

Here is what I do [on the weed volume server]

[root@media4 weed-backup]# /opt/weed/bin/weed backup -dir=. -collection=punyu.cover.images -volumeId=10 -server=XXX.YYY.ZZ:9333
I0404 12:40:18.491008 volume_loading.go:89 readSuperBlock volume 10 version 3
I0404 12:40:18.491196 volume_loading.go:136 loading index punyu.cover.images_10.idx to memory
[root@media4 weed-backup]# ls -al
total 28
drwxr-xr-x  3 weed weed 4096 Apr  4 12:28 .
drwxr-xr-x. 9 root root 4096 Oct 26 17:34 ..
-rw-r--r--  1 root root    8 Apr  4 11:29 10.dat
-rw-r--r--  1 root root    0 Apr  4 11:29 10.idx
-rwxr-xr-x  1 root root   57 Apr  4 11:29 10.vif
drwxr-xr-x  2 root root 4096 Aug  2  2022 20220801
-rw-r--r--  1 root root    8 Apr  4 12:28 punyu.cover.images_10.dat
-rw-r--r--  1 root root    0 Apr  4 12:28 punyu.cover.images_10.idx
-rwxr-xr-x  1 root root   57 Apr  4 12:28 punyu.cover.images_10.vif

Tried to do same on the Master server - with same result.

What am I doing wrong ?

Tadeusz

**Comment by tadwiew** [comment]
Just for the record - volume I am trying to backup is NOT empty. Here is weed info about it;

 volume id:10  size:349528728  collection:"punyu.cover.images"  file_count:20275  delete_count:283  deleted_byte_count:4459312  replica_placement:1  version:2  modified_at_second:1676030708

**Comment by chrislusf** [comment]
try to add some debug message in these places:
* https://github.com/seaweedfs/seaweedfs/blob/master/weed/command/backup.go#L150
* https://github.com/seaweedfs/seaweedfs/blob/master/weed/storage/volume_backup.go#L66

**Comment by tadwiew** [comment]
Hi,

1. Could not "make"weed on my CentOS 6.x server . It was failing with gcc option failure cc1: error: unrecognized command line option "-fno-lto"
    It is apparently due to very old version of gcc on CentOS 6.x 
2. "made" it on another server running AlmaLinux 9.x
     I can definitely get a file from weed volume via curl - so far so good
     BUT
    execution of new weed [tadwiew@dev9 weed-bkp]$ weed version
version 30GB 3.45  linux amd64
     ends in the same way as when using weed 3.19 - ie. one without your latest 2 modifications. No extra error msgs are generated .
[tadwiew@dev9 weed-bkp]$ weed backup  -dir=. -volumeId=10  -collection=punyu.cover.images  -server=xxx.yyy.zzz:9333
I0405 06:52:30.461129 volume_loading.go:142 loading memory index punyu.cover.images_10.idx to memory
[tadwiew@dev9 weed-bkp]$ ls -al
total 12
drwxr-xr-x. 2 tadwiew users  105 Apr  5 06:52 .
drwxr-x---. 8 tadwiew users 4096 Apr  5 06:08 ..
-rw-r--r--. 1 tadwiew users    8 Apr  5 06:52 punyu.cover.images_10.dat
-rw-r--r--. 1 tadwiew users    0 Apr  5 06:52 punyu.cover.images_10.idx
-rwxr-xr-x. 1 tadwiew users   78 Apr  5 06:52 punyu.cover.images_10.vif

**Comment by tadwiew** [comment]
Addl info.
1.  created new weed cluster on another server
2. copied  dat idx vif  files to it . Noticed that version srting is 2.
3. started master and volume
4. can get and export files
5. weed backup - still does not do anything
6. vacuum - does not do anything 
7. did export from one volume
8. created new volume, Noticed that version string is 3 [and byte offset 4]
9. uploaded to the new volume [vesrion 3] all files exported in 7. - went OK
10. run "weed backup" of that new volume [version 3]  - and it works !

So I think weed versions 3.x are not able to fully/properly handle volume files in format version 2 !
So far I see that "weed backup" and vacuum does not work for them ...

Can you please check and confirm OR provide a fix ?

**Comment by zehweh** [comment]
Hi,

today, I also tried to take backups of some of the volumes and noticed the same thing.
"Old" volumes with superblock version 2 won't work.

Whenever I try a backup of those volumes, an empty volume with superblock version 3 gets created locally but it stays empty.
I've added debug messages in the files @chrislusf mentioned above. Looks pretty identical except that "old" files immediately [get an io.EOF](https://github.com/seaweedfs/seaweedfs/blob/master/weed/storage/volume_backup.go#L89)

**Comment by chrislusf** [comment]
Correct. Older volume versions do not work. Just copy the volumes directly, since they are readonly in newer versions.

**Comment by zehweh** [comment]
They are? In the UI they show `ReadOnly: false`

In case they are readonly: Some of those "old" volumes show 30% trash - is there any way to reclaim the space?
Is there any way to "migrate" the old volumes to new superblock versions in order to have a backup method that covers all volumes?

Thanks!

**Comment by chrislusf** [comment]
to reclaim the space, use old-version `weed` binary to compact the volume.

**Comment by zehweh** [comment]
Did a bit of testing and it turns out that those volumes aren't readonly. 
They are still being written to.
Is there any way to rewrite those volumes (maybe during vacuum) to the new superblock version?

**Comment by chrislusf** [comment]
this is not supported yet.

**Comment by zehweh** [comment]
Hi,
since backing up volumes with superblock version 2 doesn't work and a rewrite is not supported yet, 
is there any other way to fetch the .dat/.vif/.idx file using `weed shell` or `wget`?
We are running seaweed in a kubernetes cluster, so `scp` is not an option.

Thanks!

**Comment by chrislusf** [comment]
@zehweh In `weed shell`, `volume.copy` can only copy to another volume server. Copying to a local directory is not supported yet unless someone adds a PR for it.

**Comment by sakethsusarla** [comment]
Can I use Restic as a tool to backup and restore data on SeaweedFS? Has anyone already tried it? The TODOs section in the [Data Backup docs page](https://github.com/seaweedfs/seaweedfs/wiki/Data-Backup#todo) of SeaweedFS suggests that restore isn't supported. I'm looking for a warm/hot backup solution for SeaweedFS.
